### DISCUSSION TITLE
Lesson 11 abi.encodePacked vs string.concat

### DISCUSSION THREAD

**Discussion by marioiordanov** [Cyfrin/foundry-full-course-cu#1321] [General]
What is the difference between using abi.encodePacked and string.concat. I tested it and both produce the same string. Also executing the test with gas-report returns the same amount of gas being used for tokenUri method

**Comment by ghost** [comment]
If you're only concatenating strings there is no difference, but `abi.encodePacked` is more convenient  because you can concatenate multiple types (address, bytesN, uint256).

With `abi.encodePacked` you can do:

```javascript
address sevenSevenSeven = address(777);
uint256 meaningOfLife = 42;
bytes memory concat = abi.encodePacked(sevenSevenSeven, meaningOfLife);
```

But the following isn't allowed:

```javascript
address sevenSevenSeven = address(777);
uint256 meaningOfLife = 42;
bytes memory concat = bytes(string.concat(sevenSevenSeven, meaningOfLife)); // won't compile because strings are required
```

The reason `string.concat` was introduced is because `abi.encodePacked` is supposed to be deprecated in favor for methods more specific, but years has passed and the method is still around.
